### PR TITLE
Issue 99 add pluggable permission support

### PR DESCRIPTION
**Pull request by thomasyip**
Issue #99

### CODE REVIEW

File: spirit/user/perm/decorators.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            })
+
+            return view_func(request, *args, **kwargs)
+        return _wrapped_view
+    if func:
+        return decorator(func)
+    return decorator
+
+def permissible_queryset(func=None, contenttype=None, or_404=True,
+            fieldlookups_kwargs={}, fieldlookups_getparams={}, fieldlookups_postparams={}, **kwargs):
```
Comment (nitely) [inline spirit/user/perm/decorators.py]: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

Comment (thomasyip) [inline spirit/user/perm/decorators.py]: Can you elaborate?

I think those params are intended to be single use.

Comment (nitely) [inline spirit/user/perm/decorators.py]: None of the function you pass those dicts modify them? Even then I'd avoid using mutables as default args.

**Comment by nitely** [comment]
```python
@permissible_object(...)
@permission_required(...)
def update(self, ...):
    pass
```

`@permissible_object` will check whether the user has  group attached to the object or not and `@permission_required` will check whether the user has the permission for that model or not, is that what's going on there?

This approach seems much more complex than [the one I proposed](https://github.com/nitely/Spirit/issues/99#issuecomment-173380626) and the decorators are kinda confusing, it actually took me a while to get what's going on and I'm not even sure I'm right. Can you explain why this is better? maybe I'm missing something.

File: spirit/settings/__init__.py [hunk 2]
```diff
@@ -6,33 +6,10 @@
...
-
-ST_ALLOWED_UPLOAD_IMAGE_FORMAT = ('jpeg', 'png', 'gif')
-
-ST_UNICODE_SLUGS = True
-
-ST_UNIQUE_EMAILS = True
-ST_CASE_INSENSITIVE_EMAILS = True
-
-ST_BASE_DIR = os.path.dirname(__file__)
+ST_BASE_DIR = os.path.join(os.path.dirname(__file__), '..')
```
Comment (nitely) [inline spirit/settings/__init__.py]: Should be os.path.dirname(os.path.dirname(__file__))

edit: no join.

File: spirit/user/perm/decorators.py (hunk 1 above)
Comment (thomasyip) [inline spirit/user/perm/decorators.py]: I am still not getting what you worry about. I think I can easily change to `None`, if it addresses your concern. Default to {} makes it readable make it clear that it is expecting a dict().

Comment (nitely) [inline spirit/user/perm/decorators.py]:
```python
def foo(asd={}, bar=True):
    if bar:
        asd['bar'] = 'bar'

    print(asd)

foo(bar=True)
# {'bar': 'bar'}
foo()
# {'bar': 'bar'}
```

It does not seem like you are mutating the dict inmediatly, but what about the functions you are passing the dict? no way to tell, even Django could be doing it internally (I doubt it, tho).

```python
def foo(asd=None, bar=True):
    asd = asd or {}
```

That easily solve it.

**Comment by thomasyip** [comment]
I am trying to stay as close as django as possible. I really don't want to maintain
my own `@permission_required(...)` in the future. But, django's does not pass the obj or query set to has_perm() test. I wanted to stay as close as possible to original syntax, which is the case now. I want to allow for model-level permission (default Django's auth backend to work with modification).

Having a `fieldlookups_kwargs` dict gives the opportunity to the auth backends to check perms with the database in single query. If you materialize it with `get_or_404(Topic, pk=pk).category` use, then has to hit the database. Without the ability to delay the materialization, I suspect each views might trigger > 2x the numbers of db queries, compare with now. Even if we don't optimize it now, I really wish the interface will allow such optimization. I think it is part of the reason that django perm is designed that way. It is designed to able to load all permission that a user has (thru `user_permissions` and `group`) in a query each, and cache it for the lifetime of the `request` object. Inventible, per-object backends will hit the database more than 2 times, but the hit should be minimized.

When using with builtin Django backends which is per-model (not per-object), we should not hit the database for topic or category at all. So, the way I implement the decorator will avoid those hit when used with builtin Django backends.

I prefer `@permission_required` over `def can_xyz` decorator. I know it is more verbose, but I can grep all @permission_required easily. But, if there is a very clear convention to put those `def can_xyz` per app, I think it would also work. (I actually looked into `askbot` code base, and they have `can_xyz` method tagged into the user object. I don't find their convention maintainable. They are also mixing db hit and code quite a bit. That makes me think `def can_xyz` should be per app.)

File: spirit/user/perm/decorators.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        all_passed = True
+        applabel_modelname, can_action = perm.split('.', 1)
+        items = _get_permissible_items(request, applabel_modelname)
+
+        if request.user.has_perms(perms):
+            if (len(self.backends) == 0):
+                return True
+            for backend in self.backends:
+                if backend.has_perms(request, perms, obj):
```
Comment (nitely) [inline spirit/user/perm/decorators.py]: Where is `obj` defined?

**Comment by nitely** [comment]
> I want to allow for model-level permission

I think you can do both model and object permission in the same query.

> If you materialize it with get_or_404(Topic, pk=pk).category use, then has to hit the database.

uhm? no. It will hit it twice, one time to check if the record exists and another to check the permission. It actually behaves correctly unlike your approach, it will raise 404 if the record does not exists and 403 if the user does not have permissions as it should be, your implementation won't.

**Comment by thomasyip** [comment]
For that view in particular, with model-level backends, category does not need to be loaded. Topic will be loaded once by `get_or_404(Topic, pk=pk)`, and your view will load it again. Without setting up some kind of cache, I think django will hit the database twice just for topic, and then for category, and then has_perm will be another query that load up all permission available for user. 4 queries.

Yeah. The implement is not optimized. So, I am just saying what "could be" done with that interface of having an optional `@permissible_object` or `@permissible_queryset`, and having @permssion_required code to call into @has_perm. The PR is not ready to merge, just to facilitate about discussion like the one we are having now. You just has to bear with me and discuss about what "could be", instead of what is.

File: spirit/user/perm/decorators.py (hunk 3 above)
Comment (thomasyip) [inline spirit/user/perm/decorators.py]: It should be `items`.

Comment (thomasyip) [inline spirit/user/perm/decorators.py]: So, it should be changed to:

```python
       if request.user.has_perms(perms):
           if (len(self.backends) == 0):
               return True

          # avoid materialize items, until we need it
           items = _get_permissible_items(request, applabel_modelname)
           for backend in self.backends:
               if backend.has_perms(request, perms, items):
```

**Comment by thomasyip** [comment]
###### I think you can do both model and object permission in the same query.

Can you elaborate on this point?

**Comment by nitely** [comment]
I edited my previous reply, I guess you missed it.

> I could actually do everything in a single query, but then I can't raise 404.
> 
> Doing the work within the view would require a single query as well, and it would raise 404 (as it does now) but I prefer the decorator, plus the object.get(pk=pk) is a fast query anyway (compared to > checking the premissions and groups which has several joins).

.

> For that view in particular, with model-level backends, category does not need to be loaded.

That's true, but if we are talking about performance, you are checking for groups and permissions in two queries instead of one and those are much more expensive than a `object.get(pk=pk)`.

> Without setting up some kind of cache

The query within the view is different, it must be done regarless, so a cache is no help here.

> I think django will hit the database twice just for topic, and then for category, and then has_perm will be another query that load up all permission available for user. 4 queries.

`get_or_404(Topic, pk=pk).category_id` will hit the database *once* (for topic only), the category pk is all that it's required. It could even be optimize it further by using `.values()` but that's something I would not bother with in a example.

The has_perm hits the database *once* to check permissions+groups, that's it.

> The PR is not ready to merge, just to facilitate about discussion like the one we are having now. You just has to bear with me and discuss about what "could be", instead of what is.

Yep, that's how open source should be, I thank you for that :smile: 

> > I think you can do both model and object permission in the same query.
> 
> Can you elaborate on this point?

`GroupPermission.objects.filter(category=category_id, category__group_permision_set__group__user_set=request.user, category__group_permision_set__permission__codename='view_topic').exists()` I think (hope) this is possible.

**Comment by thomasyip** [comment]
The permission query might not too bad usually, because those tables tend to be very few rows and stay in memory.

But, yes, if you just get id or use `values()` or `select_related()`, it could be faster.

###### GroupPermission.objects.filter(category=category_id, 

I see, you mean you don't always need a separated query for model vs category. I agree.

Let me see, so in your scenario, you are having Group and Category as 1:1 relationship. Is it something you want to support out of the box in the future? I think it is already a object-level permission (which require you to have another backends and your own decorator). 

If I don't even have category-group as object-level, I wouldn't want to join the two table though. I thought we don't have category-group right now (it is only in PR that's not merged).

    https://github.com/nitely/Spirit/blob/master/spirit/category/models.py 

So, I am assuming that you plan to merge that in?

**Comment by nitely** [comment]
Sorry, I edited again I tend to do that :smile: 

> `Category.objects.filter(pk=category_id, group_permision_set__group__user_set=request.user, group_permision_set__permission_set__codename='view_topic').exists()` I think (hope) something like this is possible.

> Or maybe `GroupPermission.objects.filter(category=category_id, group__user_set=request.user, group__permission_set__codename='view_topic').exists()`

**Comment by thomasyip** [comment]
No problem. I just wasn't sure you were talking about existing models, or something in the future. I did a quick search, and didn't see GroupPermission in the master branch (correct me if I am wrong). So, I suppose you are talking about the future.

So, say, by default, you are not having object-level permission, GroupPermission table will be empty, am I right? If it is the case, the query will be different. (sorry, if I still missed the point).

**Comment by nitely** [comment]
> I did a quick search, and didn't see GroupPermission in the master branch (correct me if I am wrong). So, I suppose you are talking about the future.

That's right.

> So, say, by default, you are not having object-level permission, GroupPermission table will be empty, am I right? If it is the case, the query will be different. (sorry, if I still missed the point).

Yes, it'll be empty, so no one will have access to the category. A more complete query would check if the category `is_private`, probably something like `GroupPermission.objects.filter(Q(group__user_set=request.user, group__permission_set__codename='view_topic') | Q(is_private=False), category=category_id).exists()`, so if not private then everyone has access, `is_private` can be setted when you create a category.

`is_private` is currently used to check if the category is the "private messages" category, for no good reason. All queries doing that right now shoud do a `pk=ST_PRIVATE_CATEGORY_PK` instead. I will fix that later myself.

**Comment by thomasyip** [comment]
It will be a pretty different discussion if you plan to have the per-object permission (adding GroupPermission) out of the box, vs staying as is.

I assume I will saved some amount of work if you have per-object permission anyway. And, my proposal might be quite different. (I had been assuming supporting current and other backends to allow for per-object).

**Comment by nitely** [comment]
Yes, [I have said so before](https://github.com/nitely/Spirit/issues/99#issuecomment-173173998).

> I assume I will saved some amount of work if you have per-object permission anyway. And, my proposal might be quite different. (I had been assuming supporting current and other backends to allow for per-object).

`GroupPermission` won't have a `ContentType`, it will be related to category only.

**Comment by thomasyip** [comment]
How soon are do you plan add that feature to master branch?

**Comment by nitely** [comment]
I can do it tomorrow probably.

I was thinking of creating a new branch, then [I saw this](https://github.com/isaacs/github/issues/18), I've never seen so many +1 haha

**Comment by nitely** [comment]
On a second thought, can you do it (since I don't know if I will merge this)? here it is:

```python
# category.models.py

# class Category(...) ...

class CategoryGroup(models.Model):

    category = models.ForeignKey('spirit_category.Category')
    group = models.ForeignKey('auth.Group')

    date = models.DateTimeField(default=timezone.now)

    class Meta:
        unique_together = ('category', 'group')
        ordering = ['-date', '-pk']
        verbose_name = _("category group")
        verbose_name_plural = _("categories groups")
```

**Comment by thomasyip** [comment]
It is my turn that I need to sleep on it. 

By having that table, you would be building a non-generic per-object backend that only work for Spirit. To have it works with my other apps, I would basically need to design my decorator such that I have an option to skip these "non-generic per-object" part of Spirit. Otherwise it, I am back to the problem that I want to avoid: two permission systems.

To make CategoryGroup generic, we would be making some kind of GroupJunctionContent, which content can be category or other model. And, it would be essentially `django-trusts` that I am building. (which is really only about 150 lines of code).

**Comment by nitely** [comment]
I thought the whole point of having a backend implementing `has_perm` and `filter_permitted_queryset` was to bypass the `CategoryGroup` and use your `MyPermission` model instead.

> To make CategoryGroup generic, we would be making some kind of GroupJunctionContent, which content can be category or other model. And, it would be essentially django-trusts that I am building. (which is really only about 150 lines of code).

The backend is were I trace the limit haha.

**Comment by thomasyip** [comment]
##### the whole point of having a backend implementing has_perm and filter_permitted_queryset was to bypass the CategoryGroup

Maybe it isn't that different. But, I set my mind to support both model-level and object-level permission with Spirit, until two hours ago. 

It might actually be easier to support two slightly different object-level permissions.

**Comment by nitely** [comment]
> It might actually be easier to support two slightly different object-level permissions.

You can probably check within your form if the permission belongs to `Category` and create the group-category manually. Then there would be no backend required. But I don't know the inners of your project, so it is just a thought.

**Comment by thomasyip** [comment]
Some app in my project is built entirely with REST and ajax. So, I don't have "django form" for them. (the Spirit part won't be REST or ajax, because the need to have a forum inside an iOS / Android app not very strong).

I feel that I should still find a way to manage Category as just another "content". It will be easier down the road.

File: spirit/user/perm/decorators.py (hunk 1 above)
Comment (thomasyip) [inline spirit/user/perm/decorators.py]: O, I see, "the same list is used in each successive call".

I certainly misunderstood it. Thanks for pointing it out.

**Comment by thomasyip** [comment]
That's actually won't work to define permissible_object() and permission_required seperately.

I won't able to tell which is which. And, if one permissible_object of results is NotExist, I won't know how to handle it.

```python
@permissible_object('spirit_category', fieldlookups_kwargs={'category_topic_pk': 'topic_pk'})
@permissible_object('spirit_category', fieldlookups_kwargs={'pk': 'category_pk_moving_to'})
@permission_required('spirit_category.can_add')
@permission_required('spirit_category.can_read')
``

**Comment by thomasyip** [comment]
So, I think I am getting to the bottom of it. I am understanding much better on what you already told me.

In short, I don't think there is a very feasible to make a generic backend that would work with different object-level backends. Perhaps, we can make two specific backend to work, but each one would have some assumptions and we would add more method to the backends.

For example, there are many "features" that a specific backend might implement: Roles, Inheritance, Logical checks, Cascade update, Model-level. Those features can make sense for Spirit. But if it is used, it would be possible of support those doesn't as Spirit's authentication backends.

So, I have made my changes with django-trust. I got to a point that I flushed out what needed to be built into django-trust, and believe I can make Spirit to work for my permission use-cases.

I didn't think you were ready to pin Spirit to a specific permission backend. 

I still push it up anyway as a knowledge sharing. But, I am not going to make PR merge-able.

**Pull request closed without merging**